spark3/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanUtils.scala

```diff
@@ -54,14 +55,28 @@ object PlanUtils {
         Integer.TYPE)
       .build()
 
+  def numShufflePartitions(conf: SQLConf): Option[Int] = {
+    if (Spark3VersionUtil.isSpark30) {
+      Some(conf.numShufflePartitions)
+    } else {
+      // SPARK-32056: Coalesce partitions for repartition by expressions when AQE is enabled
+      None
```

Comment:
Do we need to check if AQE is enabled? E.g. if a user disables AQE, should this still return `Some(conf.numShufflePartitions)`?

Reply:
We don't need to do it, see https://github.com/apache/spark/pull/28900/files#diff-0ac2c89f8cc0d00e8fe7717b01697f36f20fe8abf2def09bcfde0ad84b30e467R968